src/pool.ts

```diff
@@ -0,0 +1,677 @@
+      // Connections have been expired, re-connect the holder.
+      await this._connection.close();
+      this._connection = null;
+      await this.connect();
+    }
+
+    const proxy = new PoolConnectionProxy(this, this.getConnectionOrThrow());
+    this._proxy = proxy;
+
+    if (this._onAcquire) {
+      try {
+        await this._onAcquire(proxy);
+      } catch (error) {
+        // If a user-defined `onAcquire` function fails, we don't
+        // know if the connection is safe for re-use, hence
+        // we close it.  A new connection will be created
+        // when `acquire` is called again.
+
+        // Use `close()` to close the connection gracefully.
+        // An exception in `onAcquire` isn't necessarily caused
+        // by an IO or a protocol error.  close() will
+        // do the necessary cleanup via _release_on_close().
+
+        // TODO: check _release_on_close in Connection
+        await this._connection?.close();
+        throw error;
+      }
+    }
+
+    this._inUse = new Deferred<void>();
+    return proxy;
+  }
+
+  async release(): Promise<void> {
+    if (this._inUse === null) {
+      throw new InternalClientError(
+        "PoolConnectionHolder.release() called on " +
+          "a free connection holder"
+      );
+    }
+
+    if (this._connection?.isClosed()) {
+      // When closing, pool connections perform the necessary
+      // cleanup, so we don't have to do anything else here.
+      return;
+    }
+
+    if (this._generation !== this._pool._generation) {
+      // The connection has expired because it belongs to
+      // an older generation (AsyncIOPool.expire_connections() has
+      // been called).
+      await this._connection?.close();
+      return;
+    }
+
+    if (this._onRelease && this._proxy) {
+      try {
+        await this._onRelease(this._proxy);
+      } catch (error) {
+        // If a user-defined `onRelease` function fails, we don't
+        // know if the connection is safe for re-use, hence
+        // we close it.  A new connection will be created
+        // when `acquire` is called again.
+        await this._connection?.close();
+        // Use `close()` to close the connection gracefully.
+        // An exception in `setup` isn't necessarily caused
+        // by an IO or a protocol error.  close() will
+        // do the necessary cleanup via _release_on_close().
+        throw error;
+      }
+    }
+
+    // Free this connection holder and invalidate the
+    // connection proxy.
+    this._release();
+  }
+
+  async waitUntilReleased(): Promise<void> {
+    if (this._inUse === null) {
+      return;
+    }
+    await this._inUse.promise;
+  }
+
+  async close(): Promise<void> {
+    if (this._connection !== null) {
+      // TODO: complete, remove following comment
+      // AsyncIOConnection.aclose() will call _release_on_close() to
+      // finish holder cleanup.
+      await this._connection.close();
+    }
+  }
+
+  _release_on_close(): void {
```

Comment:
Let's camelCase everywhere.

Reply:
An oversight. By the way, how do you feel about private methods starting with an underscore? I have mixed feelings about this, because TypeScript supports `private` anyway.